### ISSUE TITLE
The masking matrix seems to be wrong in the "torch.nn.functional.scaled_dot_product_attention" function

### ISSUE THREAD

**Issue by DongHande** [pytorch/pytorch#110144]
### 🐛 Describe the bug

It seems to be a logic error about the flash attention in transformer model in the implementation of "torch.nn.functional.scaled_dot_product_attention" function (named SDPA below). *But I am not sure because current implementation may consider other situations. This problem can be solved by specify the attn_mask parameter in SDPA function. But if it is a logic error, maybe it should be fixed for more convenient to use.* 

Situation: I use the GPT to generate text with **the KV-cache** and **more than more tokens**.  For example, I have the KV-cache of the first 4 tokens (t_0, t_1, t_2, t_3) denoted as K[:4] and V[:4], and have two new tokens (t_4, t_5) to infer the 7-th token. With the (t_4, t_5) tokens, I can calculate their QKV denoted as Q[4:6], K[4:6] and V[4:6]. 

Correct procession: We should concatenate the KVs of above tokens and current tokens as K[:6] and V[:6], and next conduct the masking attention about Q[4:6], K[:6] and V[:6] with a masking matrix [[True, True, True, True, True, False], [True, True, True, True, True, True]]. 

Problem: I read the document, and it seems that the masking matrix in the SDPA implemention is [[True, False, False, False, False, False], [True, True, False, False, False, False]], resulting wrong result. 

Below it my code to verify my guess: 
```
import torch
query = torch.randn(1, 1, 2, 3).to("cuda") # (batch_size, num_heads, query_len, dim)
key = torch.randn(1, 1, 6, 3).to("cuda") # (batch_size, num_heads, key_len, dim)
value = torch.randn(1, 1, 6, 3).to("cuda") # (batch_size, num_heads, value_len, dim)

# with all KV cache
torch.nn.functional.scaled_dot_product_attention(
            query, key, value, attn_mask=None, dropout_p=0.0, is_causal=True
        )
-----------------------
tensor([[[[ 0.8399,  0.2988,  0.6063],
          [ 0.0390, -0.3336,  0.3737]]]], device='cuda:0')


# With only the first two KV-cache
# If my guess is right, it should be the same with previous tensor because the masking matrix results in the below KV not working. 
torch.nn.functional.scaled_dot_product_attention(
            query, key[:,:,:2,:], value[:,:,:2,:], attn_mask=None, dropout_p=0.0, is_causal=True
        )
----------------------
tensor([[[[ 0.8399,  0.2988,  0.6063],
          [ 0.0390, -0.3336,  0.3737]]]], device='cuda:0')

# Desired operation without flash attention
attention_mask = torch.tensor([[[True, True, True, True, True, False], [True, True, True, True, True, True]]]).to("cuda")
attention_mask = attention_mask.unsqueeze(1)
attention_mask = torch.where(attention_mask, 0.0, -3.3895e+38)

attn_weights = query @ key.transpose(-2, -1) / query.size(-1)**0.5
attn_weights = attn_weights + attention_mask
attn_weights = torch.nn.functional.softmax(attn_weights, dim=-1)
attn_weights @ value
-------------------------
tensor([[[[0.2244, 0.1375, 0.2416],
          [0.0936, 0.1238, 0.8702]]]], device='cuda:0')

# with right attn_mask parameter, it works well. 
# But if current implementation is not reasonable, maybe it can be fixed for more convenient use. 

attention_mask = torch.tensor([[[True, True, True, True, True, False], [True, True, True, True, True, True]]]).to("cuda")
attention_mask = attention_mask.unsqueeze(1)
attention_mask = torch.where(attention_mask, 0.0, -3.3895e+38)
torch.nn.functional.scaled_dot_product_attention(
            query, key, value, attn_mask=attention_mask, dropout_p=0.0, is_causal=False
        )
--------------------------
tensor([[[[0.2244, 0.1375, 0.2416],
          [0.0936, 0.1238, 0.8702]]]], device='cuda:0')
```

### Versions

Collecting environment information...
PyTorch version: 2.0.1+cu118
Is debug build: False
CUDA used to build PyTorch: 11.8
ROCM used to build PyTorch: N/A

OS: TencentOS Server 3.1 (Final) (x86_64)
GCC version: (GCC) 8.5.0 20210514 (TencentOS 8.5.0-18)
Clang version: 13.0.1 (Red Hat 13.0.1-2.module+el8.6.0+37+eac49f58)
CMake version: version 3.25.0
Libc version: glibc-2.28

Python version: 3.10.13 (main, Sep 11 2023, 13:44:35) [GCC 11.2.0] (64-bit runtime)
Python platform: Linux-5.4.119-19-0009.11-x86_64-with-glibc2.28
Is CUDA available: True
CUDA runtime version: 11.8.89
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: 
GPU 0: NVIDIA A10
GPU 1: NVIDIA A10
GPU 2: NVIDIA A10
GPU 3: NVIDIA A10

Nvidia driver version: Could not collect
cuDNN version: Probably one of the following:
/usr/local/cuda-11.4/targets/x86_64-linux/lib/libcudnn.so.8
/usr/local/cuda-11.4/targets/x86_64-linux/lib/libcudnn_adv_infer.so.8
/usr/local/cuda-11.4/targets/x86_64-linux/lib/libcudnn_adv_train.so.8
/usr/local/cuda-11.4/targets/x86_64-linux/lib/libcudnn_cnn_infer.so.8
/usr/local/cuda-11.4/targets/x86_64-linux/lib/libcudnn_cnn_train.so.8
/usr/local/cuda-11.4/targets/x86_64-linux/lib/libcudnn_ops_infer.so.8
/usr/local/cuda-11.4/targets/x86_64-linux/lib/libcudnn_ops_train.so.8
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:        x86_64
CPU op-mode(s):      32-bit, 64-bit
Byte Order:          Little Endian
CPU(s):              112
On-line CPU(s) list: 0-111
Thread(s) per core:  2
Core(s) per socket:  56
Socket(s):           1
NUMA node(s):        1
Vendor ID:           AuthenticAMD
BIOS Vendor ID:      Smdbmds
CPU family:          25
Model:               1
Model name:          AMD EPYC 7K83 64-Core Processor
BIOS Model name:     3.0
Stepping:            0
CPU MHz:             2545.218
BogoMIPS:            5090.43
Hypervisor vendor:   KVM
Virtualization type: full
L1d cache:           32K
L1i cache:           32K
L2 cache:            512K
L3 cache:            32768K
NUMA node0 CPU(s):   0-111
Flags:               fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ht syscall nx mmxext fxsr_opt pdpe1gb rdtscp lm constant_tsc rep_good nopl nonstop_tsc cpuid extd_apicid tsc_known_freq pni pclmulqdq monitor ssse3 fma cx16 sse4_1 sse4_2 x2apic movbe popcnt aes xsave avx f16c rdrand hypervisor lahf_lm cmp_legacy cr8_legacy abm sse4a misalignsse 3dnowprefetch osvw topoext ibpb vmmcall fsgsbase bmi1 avx2 smep bmi2 erms rdseed adx smap clflushopt sha_ni xsaveopt xsavec xgetbv1 arat fsrm

Versions of relevant libraries:
[pip3] numpy==1.24.1
[pip3] torch==2.0.1+cu118
[pip3] torchaudio==2.0.2+cu118
[pip3] torchvision==0.15.2+cu118
[pip3] triton==2.0.0
[conda] numpy                     1.24.1                   pypi_0    pypi
[conda] torch                     2.0.1+cu118              pypi_0    pypi
[conda] torchaudio                2.0.2+cu118              pypi_0    pypi
[conda] torchvision               0.15.2+cu118             pypi_0    pypi
[conda] triton                    2.0.0                    pypi_0    pypi

cc @jbschlosser @bhosmer @cpuhrsch @erichan1 @drisspg

**Comment by drisspg** [comment]
I would argue that the masking matrix is not "wrong" per say rather it chose between two patterns. For more discussion see this issue: 
https://github.com/pytorch/pytorch/issues/108108

**Comment by DongHande** [comment]
Thanks for your clarification! The explanation in your mentioned issue is quite clear. 

I will set the attn_mask parameter to set the masking matrix for my use.
